File: app/code/parents.py
class Person:
    def __init__(self, name="", father=None, mother=None):
        self.name = name
        self.father = father
        self.mother = mother

def parents(person):
    result = []
    if person is not None:
        if person.father is not None:
            result.append(person.father)
        if person.mother is not None:
            result.append(person.mother)
    return result

def grandparents(person):
    result = []
    for p in parents(person):
        p = parents(p)
        if p is not None:
            result.append(p)
    return result

File: app/code/parents.java
public class Person {
    private String name;
    private Person father;
    private Person mother;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Person getFather() {
        return father;
    }

    public void setFather(Person father) {
        this.father = father;
    }

    public Person getMother() {
        return mother;
    }

    public void setMother(Person mother) {
        this.mother = mother;
    }

    public List<Person> getParents() {
        List<Person> parents = new List<Person>();
        if (this.father != null) {
            parents.add(this.father);
        }
        if (this.mother != null) {
            parents.add(this.mother);
        }
        return parents;
    }

    public List<Person> getGrandparents() {
        List<Person> grandparents = new List<Person>();
        for (Person p : getParents()) {
            grandparents.addAll(p.getParents());
        }
        return grandparents;
    }
}
